tpls/desul/include/desul/atomics/Compare_Exchange_OpenACC.hpp

```diff
@@ -0,0 +1,200 @@
+  static_assert(sizeof(unsigned int) == 4,
+                "this function assumes an unsigned int is 32-bit");
+  unsigned int return_val = atomicCAS(reinterpret_cast<unsigned int*>(dest),
+                                      reinterpret_cast<unsigned int&>(compare),
+                                      reinterpret_cast<unsigned int&>(value));
+  return reinterpret_cast<T&>(return_val);
+}
+
+#pragma acc routine seq
+template <class T, class MemoryScope>
+std::enable_if_t<std::is_integral<T>::value && (sizeof(T) == 8) \
+	&& (std::is_same_v<MemoryScope,MemoryScopeDevice>           \
+	|| std::is_same_v<MemoryScope,MemoryScopeCore>), T>
+device_atomic_compare_exchange(T* const dest, T compare, T value, MemoryOrderRelaxed, MemoryScope) {
+  static_assert(sizeof(unsigned long long int) == 8,
+                "this function assumes an unsigned long long int is 64-bit");
+  unsigned long long int return_val = atomicCAS(reinterpret_cast<unsigned long long int*>(dest),
+                                      reinterpret_cast<unsigned long long int&>(compare),
+                                      reinterpret_cast<unsigned long long int&>(value));
+  return reinterpret_cast<T&>(return_val);
+}
+
+#pragma acc routine seq
+template <class T, class MemoryScope>
+std::enable_if_t<std::is_floating_point<T>::value && (sizeof(T) == 4) \
+	&& (std::is_same_v<MemoryScope,MemoryScopeDevice>                 \
+	|| std::is_same_v<MemoryScope,MemoryScopeCore>), T>
+device_atomic_compare_exchange(T* const dest, T compare, T value, MemoryOrderRelaxed, MemoryScope) {
+  static_assert(sizeof(float) == 4,
+                "this function assumes an float is 32-bit");
+  float return_val = atomicCAS(reinterpret_cast<float*>(dest),
+                                      reinterpret_cast<float&>(compare),
+                                      reinterpret_cast<float&>(value));
+  return reinterpret_cast<T&>(return_val);
+}
+
+#ifndef DESUL_CUDA_ARCH_IS_PRE_PASCAL
+
+#pragma acc routine seq
+template <class T, class MemoryScope>
+std::enable_if_t<std::is_floating_point<T>::value && (sizeof(T) == 8) \
+	&& (std::is_same_v<MemoryScope,MemoryScopeDevice>                 \
+	|| std::is_same_v<MemoryScope,MemoryScopeCore>), T>
+device_atomic_compare_exchange(T* const dest, T compare, T value, MemoryOrderRelaxed, MemoryScope) {
+  static_assert(sizeof(double) == 8,
+                "this function assumes an double is 64-bit");
+  double return_val = atomicCAS(reinterpret_cast<double*>(dest),
+                                      reinterpret_cast<double&>(compare),
+                                      reinterpret_cast<double&>(value));
+  return reinterpret_cast<T&>(return_val);
+}
+
+#endif
+
+#else
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+T device_atomic_compare_exchange(T* dest, T compare, T value, MemoryOrder, MemoryScope scope) {
```

Comment:
The current OpenACC stand does not support compare-and-exchange atomic operations at all; that's why CUDA intrinsics are used for the NVHPC case. 
For the LLVM-Clacc case, I have not yet checked what kind of alternative options are available. One possibility would be to use OpenMP-specific implementation for the Clacc case since Clacc itself is built on top of LLVM OpenMP implementation. I will dig into this direction in the next PR after this PR is merged.